File: Java/DSA/Recursion/Power.java
package Java.DSA.Recursion;

import java.math.BigInteger;
import java.util.*;

public class Power {
    public static BigInteger pow(BigInteger n, int p){
        if(p == 1) return n;

        if(p%2 != 0){
            return n.multiply(pow(n, p - 1));
        }

        return pow(n, p/2).multiply(pow(n, p/2));
    }

    public static void main(String[] args) {

        BigInteger a;
        int p;
        Scanner sc = new Scanner(System.in);
        a = sc.nextBigInteger();
        p = sc.nextInt();
        BigInteger c = pow(a, p);
        System.out.println(c);
        sc.close();
    
    }
}

File: Recursion & Backtracking/Power.cpp
#include <iostream>
using namespace std;

int power(int n, int p){
    if(p == 1)
    return n;

    if(n == 1)
    return n;

    //return n * power(n, p - 1); T = O(n)
    return power(n, p/2) * power(n, p/2); //  T = O(log n)
}

int main(){
    cout << power(1, 3);
    return 0;
}